staging/src/k8s.io/client-go/testing/fixture.go

```diff
@@ -367,6 +368,19 @@ func (t *tracker) add(gvr schema.GroupVersionResource, obj runtime.Object, ns st
 	// copy.
 	obj = obj.DeepCopyObject()
 
+	if unstruct, ok := obj.(*unstructured.Unstructured); ok {
+		// Make a best-effort to translate the provided Unstructured object to a concrete type. If successful, then
+		// strategic patches will work on this instance, otherwise they will fail.
+		kinds, _, err := t.scheme.ObjectKinds(obj)
+		if err == nil && len(kinds) > 0 {
+			if newObj, err := t.scheme.New(kinds[0]); err == nil {
+				if err = runtime.DefaultUnstructuredConverter.FromUnstructured(unstruct.Object, newObj); err == nil {
```

Comment:
it should be only working for patch, or odds things will happen like, if you create w/ an unstructured but get a typed instance returned, which is, breaking more essential behavior of the fake client. additionally, i think the conversion can't mock out apiserver's behavior, encoding by unstructured-scheme and then decoding using the provided scheme will be better.

Reply:
I checked the fake client code with the following helper program:

```golang
	var unst unstructured.Unstructured
	unst.UnmarshalJSON([]byte(`{"apiVersion": "v1", "kind": "Secret", "type": "Opaque", "metadata": {"name": "test", "namespace": "testns"}, "data": {"test": "dGVzdAo="} }`))

	client.CoreV1().Namespaces().Create(&v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "testns"}})
	dynSecret := dyn.Resource(schema.GroupVersionResource{Group: "", Version: "v1", Resource: "secrets"})
	if _, e := dynSecret.Namespace("testns").Create(&unst, metav1.CreateOptions{}); e != nil {
		panic(e)
	}
	if us, e := dynSecret.Namespace("testns").Get("test", metav1.GetOptions{}); e != nil {
		panic(e)
	} else {
  	       fmt.Printf("Unstructured Secret is %#v\n", us)
        }
	if s, e := client.CoreV1().Secrets("testns").Get("test", metav1.GetOptions{}); e != nil {
		panic(e)
	} else {
	     fmt.Printf("Secret is %#v\n", s)
        }
```
The original fake client (before and after my suggested patch) crashes when trying to retrieve the secret as a `v1.Secret` object. I think that this is a different bug altogether.

Looking at the API provided by the client, if you use the typed clientset to retrieve an instance, then the fake client must return a typed instance, regardless of how that instance was created. This is exactly how the API is defined, so doing anything else is a bug - the fake client wouldn't behave like the thing it's trying to fake. I don't think it's essential behavior for the fake client to support a flow that doesn't work with the real apiserver.

The reverse is also impossible: regardless of how you created the object (with a typed instance or an Unstructured one), if you use the dynamic client you expect an `Unstructured` object to be returned. The code even does this translation in the dynamic simple client.

I looked at doing the translation only before the strategic patch, but the code doesn't have access to the ObjectScheme there, so I couldn't actually perform the conversion.

I'm new to the Kubernetes object model - I don't follow what you mean by encoding by the Unstructured-scheme and decoding with the provided scheme. Could you show me an example of this, or a link to the correct API?